File: implementations/java-spring/Application.java
package com.example.gateway;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@SpringBootApplication
public class Application {
  public static void main(String[] args) {
    SpringApplication.run(Application.class, args);
  }

  @RestController
  @RequestMapping("/api/v1")
  static class ApiController {
    private final Map<String, User> users = new ConcurrentHashMap<>();
    private final Map<String, String> access = new ConcurrentHashMap<>();
    private final Map<String, String> refresh = new ConcurrentHashMap<>();
    private volatile long uptime = 0;

    @PostMapping("/auth/register")
    public ResponseEntity<?> register(@RequestBody Map<String, String> body) {
      String username = body.get("username");
      String password = body.get("password");
      String email = body.getOrDefault("email", username + "@example.local");
      if (username == null || password == null) return ResponseEntity.badRequest().build();
      if (users.containsKey(username)) return ResponseEntity.status(HttpStatus.CONFLICT).body(Map.of("error", Map.of("code", "CONFLICT")));
      users.put(username, new User(username, email, password, "viewer"));
      return ResponseEntity.status(HttpStatus.CREATED).body(Map.of("status", "registered"));
    }

    @PostMapping("/auth/login")
    public ResponseEntity<?> login(@RequestBody Map<String, String> body) {
      User user = users.get(body.get("username"));
      if (user == null || !user.password().equals(body.get("password"))) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
      String at = UUID.randomUUID().toString().replace("-", "");
      String rt = UUID.randomUUID().toString().replace("-", "");
      access.put(at, user.username());
      refresh.put(rt, user.username());
      return ResponseEntity.ok(Map.of("accessToken", at, "refreshToken", rt));
    }

    @PostMapping("/auth/refresh")
    public ResponseEntity<?> refresh(@RequestBody Map<String, String> body) {
      String username = refresh.get(body.get("refreshToken"));
      if (username == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
      String at = UUID.randomUUID().toString().replace("-", "");
      access.put(at, username);
      return ResponseEntity.ok(Map.of("accessToken", at, "refreshToken", body.get("refreshToken")));
    }

    @PostMapping("/auth/logout")
    public ResponseEntity<?> logout(@RequestBody Map<String, String> body, @RequestHeader("Authorization") String auth) {
      if (currentUser(auth) == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
      refresh.remove(body.get("refreshToken"));
      return ResponseEntity.ok(Map.of("status", "ok"));
    }

    @GetMapping("/auth/me")
    public ResponseEntity<?> me(@RequestHeader("Authorization") String auth) {
      User user = currentUser(auth);
      if (user == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
      return ResponseEntity.ok(Map.of("username", user.username(), "email", user.email(), "role", user.role()));
    }

    @GetMapping("/metrics/current")
    public ResponseEntity<?> metrics(@RequestHeader("Authorization") String auth){
      if (currentUser(auth) == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
      uptime++;
      return ResponseEntity.ok(Map.of("temperatureC",25.0,"cpuLoadPercent",20.0,"ramLoadPercent",40.0,"uptimeSeconds",uptime,"supplyVoltageV",12.2,"timestampUtc", Instant.now().toString()));
    }

    @GetMapping("/gateway/status")
    public Map<String,Object> status(){ return Map.of("service","ok","opcua","simulated","cache","ready"); }

    private User currentUser(String authorization) {
      if (authorization == null || !authorization.startsWith("Bearer ")) return null;
      String token = authorization.substring("Bearer ".length());
      String username = access.get(token);
      return username == null ? null : users.get(username);
    }
  }

  record User(String username, String email, String password, String role) {}
}

File: implementations/kotlin-spring/Application.kt
package com.example.gateway

import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@SpringBootApplication
class Application
fun main(args: Array<String>) = runApplication<Application>(*args)

data class User(val username: String, val email: String, val password: String, val role: String = "viewer")

@RestController
@RequestMapping("/api/v1")
class ApiController {
    private val users = ConcurrentHashMap<String, User>()
    private val access = ConcurrentHashMap<String, String>()
    private val refresh = ConcurrentHashMap<String, String>()
    private var uptime: Long = 0

    @PostMapping("/auth/register")
    fun register(@RequestBody body: Map<String, String>): ResponseEntity<Any> {
        val username = body["username"] ?: return ResponseEntity.badRequest().build()
        val password = body["password"] ?: return ResponseEntity.badRequest().build()
        if (users.containsKey(username)) return ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("error" to mapOf("code" to "CONFLICT")))
        users[username] = User(username, body["email"] ?: "$username@example.local", password)
        return ResponseEntity.status(HttpStatus.CREATED).body(mapOf("status" to "registered"))
    }

    @PostMapping("/auth/login")
    fun login(@RequestBody body: Map<String, String>): ResponseEntity<Any> {
        val user = users[body["username"]] ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        if (user.password != body["password"]) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        val at = UUID.randomUUID().toString().replace("-", "")
        val rt = UUID.randomUUID().toString().replace("-", "")
        access[at] = user.username
        refresh[rt] = user.username
        return ResponseEntity.ok(mapOf("accessToken" to at, "refreshToken" to rt))
    }

    @PostMapping("/auth/refresh")
    fun refreshToken(@RequestBody body: Map<String, String>): ResponseEntity<Any> {
        val username = refresh[body["refreshToken"]] ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        val at = UUID.randomUUID().toString().replace("-", "")
        access[at] = username
        return ResponseEntity.ok(mapOf("accessToken" to at, "refreshToken" to body["refreshToken"]))
    }

    @PostMapping("/auth/logout")
    fun logout(@RequestBody body: Map<String, String>, @RequestHeader("Authorization") auth: String?): ResponseEntity<Any> {
        if (currentUser(auth) == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        refresh.remove(body["refreshToken"])
        return ResponseEntity.ok(mapOf("status" to "ok"))
    }

    @GetMapping("/auth/me")
    fun me(@RequestHeader("Authorization") auth: String?): ResponseEntity<Any> {
        val user = currentUser(auth) ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        return ResponseEntity.ok(mapOf("username" to user.username, "email" to user.email, "role" to user.role))
    }

    @GetMapping("/metrics/current")
    fun metrics(@RequestHeader("Authorization") auth: String?): ResponseEntity<Any> {
        if (currentUser(auth) == null) return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build()
        uptime += 1
        return ResponseEntity.ok(mapOf("temperatureC" to 25.0, "cpuLoadPercent" to 22.0, "ramLoadPercent" to 51.0, "uptimeSeconds" to uptime, "supplyVoltageV" to 12.1, "timestampUtc" to Instant.now().toString()))
    }

    @GetMapping("/gateway/status")
    fun status() = mapOf("service" to "ok", "opcua" to "simulated", "cache" to "ready")

    private fun currentUser(auth: String?): User? {
        if (auth == null || !auth.startsWith("Bearer ")) return null
        return users[access[auth.removePrefix("Bearer ")] ?: return null]
    }
}
